### DISCUSSION TITLE
Kernel Status seems to represent both the server as well as the client states for the kernel

### DISCUSSION THREAD

**Discussion by shibbas** [nteract/archived-desktop-app#5552] [Q&A]
It looks like the kernel.status stored in the redux store is used both to represent the client action (Restarting | Starting) for the kernels as well as the actual status reported by the kernel. Reference https://github.com/nteract/nteract/blob/df958a94565520df1645b54ba7218b24be4e1a6a/packages/reducers/src/core/entities/kernels.ts#L51 

Was wondering if it was better to represent these states separately or have kernel status only represent what the server returns over the channel or on launch and have a separate entity to track the client action? The reason is because it feels like one would mask the other. Some scenarios I can think of: 

- Connecting to an new kernel. The kernel.status is set "Starting" before the kernel is even started on the server. 
- Connecting to an existing kernel. The kernel.status is set to "Starting" whereas the kernel is actually already started and had probably been active on the server until the launch API returns "Idle". 
- Restarting a kernel. Restarting will set the status to "Restarting" which can take a bit of time and during that time, the kernel is still processing. So the actual state of the kernel on the server is lost. 

Thoughts?

**Comment by willingc** [comment]
Actions can come from:

Primary Client

Peer Client

Kernel Server

**Comment by willingc** [comment]
Identify State Transitions and Actions/Expected Behavior

**Comment by MSeal** [comment]
So the complication here is that Jupyter traditionally thinks of the browser as the source of truth. This means there's multiple lines of communication about actions taken.

Keeping the UI messaging similar to Jupyter Classic is probably a good default for the final status. This means the final state should probably match into the redux store but different actions should be isolated.

The primary issue I believe will be around the fact that `execute_reply` can be out of order with idle -- you need to wait for either idle or reply rather than just `execute_reply` in cell state transitions. But these messages can come out of order so you need to do A) if `execute_reply` comes back set state and B) if `idle` state comes back wait 1 second for the `execute_reply` before giving up on it to transition state and assume browser state was invalidated.
